### ISSUE TITLE
Files that incorrectly have the .jpg/jpeg extension cause editor/instance to crash (on 64bit?)

### ISSUE THREAD

**Issue by Goblinth** [goostengine/goost#83]
**Goost and Godot version:**
Goost windows v1.0 stable

**OS/device including version:**
Windows 10
Version 1909

**Issue description:**
Attempting to import (through editor or code) a non-jpeg file that has the .jpg/jpeg extension will cause the editor/running instance to crash.

This happens on these versions:
Godot_v3.3.2-stable.goost_v1.0-stable_mono_win64.zip
Godot_v3.3.2-stable.goost_v1.0-stable_win64.exe.zip

Whereas on these versions it will correctly display a red X for the icon / fail with an error:
Godot_v3.3.2-stable.goost_v1.0-stable_win32.exe.zip
Godot_v3.3.2-stable.goost_v1.0-stable_mono_win32.zip

Note that I tested the other supported formats and it seems only the .jpeg/jpg extension causes this to happen.

**Steps to reproduce:**
Change the extension of a file that is not a jpeg to .jpg/.jpeg 
Try to import it in the editor.
Or try to load it in a script with Image.load()

**Minimal reproduction project:**
If the import attempt was done in the editor instead of code, then the project will fail to start until the offending file(s) are removed from its directory. Here is a (not) jpg for testing (a text file that says TEST) :
![ttt](https://user-images.githubusercontent.com/85024455/120078325-7aa60b00-c09e-11eb-8912-2c88c7eb22cb.jpg)

[jpeg-crash.zip](https://github.com/goostengine/goost/files/6565104/jpeg-crash.zip)

**Comment by Xrayez** [comment]
That's interesting. 😄

I confirm that this only happens with Goost binaries on 64-bit OS (also Windows).

It's very strange that this would cause a crash in Goost binaries but not Godot ones. I haven't yet debugged the problem but this could be caused by:

- Goost interfering with import process. I think it's possible, but I haven't touched JPEG import code in Goost at all.
- Using different buildsystem to build Godot from source, which may expose a hidden bug (possibly the code which has undefined behavior) unreachable via normal means. I'm using official Godot build scripts and containers but yeah I did have issues compiling the binaries on my Fedora 34 host. In that case, it's something which could be fixed in Godot.
- Perhaps optimizations like LTO could make the code slightly different compared to official Godot builds.

The good news is that valid JPEG files do import and load correctly.

**Comment by Xrayez** [comment]
The crash does *not* happen in `target=debug` nor `target=release_debug` builds (MSVC compiler):

![image](https://user-images.githubusercontent.com/17108460/120080504-f49bbb80-c0c1-11eb-8148-963feea28fe2.png)

This suggests that it may be due to optimization differences in Godot vs Goost. But I do not apply any custom compiler optimization options and try to replicate Godot build process as close as possible. It's also possible that recent versions of Godot (`3.4-beta`) could already have fixed this issue, but that sounds unlikely.

The crash may happen using MinGW, that both official Godot buildsystem and Goost use for compiling Windows builds. So, I wonder if this is specific to Windows/MinGW-based compiler.

- [x] `scons target=release_debug use_lto=yes` - MSVC, does not crash.
- [x] `scons target=release_debug use_mingw=yes` - MinGW, does not crash on Fedora.
- [x] `scons target=release_debug use_lto=yes use_mingw=yes` - Crash on Fedora, see https://github.com/goostengine/goost/issues/83#issuecomment-851067288.

**Comment by Goblinth** [comment]
I'm not able to compile to test. If I compile with use_mingw=yes I get the following error:
(Adding the other options to it cause the same error)

_[Initial build] Compiling ==> platform\windows\godot_windows.cpp
scons: *** [platform\windows\godot_windows.windows.opt.tools.64.goost.o] The system cannot find the file specified
scons: building terminated because of errors._

If I try to compile without use_mingw=yes I get this error instead:

_[Initial build] Compiling ==> platform\windows\os_windows.cpp
os_windows.cpp
W:\_TTTMTMPP\goost\godot\core/math/bvh_abb.h(107): warning C4346: 'BVH_ABB<Bounds,Point>::ConvexHull': dependent name is not a type
W:\_TTTMTMPP\goost\godot\core/math/bvh_abb.h(107): note: prefix with 'typename' to indicate a type
W:\_TTTMTMPP\goost\godot\core/math/bvh_abb.h(274): note: see reference to class template instantiation 'BVH_ABB<Bounds,Point>' being compiled
W:\_TTTMTMPP\goost\godot\core/math/bvh_abb.h(107): error C2061: syntax error: identifier 'ConvexHull'
scons: *** [platform\windows\os_windows.windows.tools.64.goost.obj] Error 2
scons: building terminated because of errors._

**Comment by Xrayez** [comment]
You may need to update Visual Studio compiler in order to compile I guess (using VS 2019).

I've tested this with `scons target=release_debug use_lto=yes` and this does not crash either. So, without actually testing to confirm this yet, I guess it's indeed MinGW-specific issue with LTO.

Godot has similar reports for EXR image format: godotengine/godot#24641. They seem to have fixed it by fixing/updating MinGW toolchain (both upstream and locally). I could contact Godot dev team to get some insights on this, or at least confirm the hypothesis that this could be the culprit.

Likely JPEG loader could also be fixed in Godot, there's nothing I can do from the Goost side to fix it (on the level of the source code), or rather it should probably be fixed in Godot. But I also wouldn't want to fix it on the level of the buildsystem (I have little to no expertise on compiler development, nor want to maintain custom buildsystem patches to fix the issue, as it would be quite PITA). 😅

So, I'm not yet sure what needs to be done to fix it.

**Comment by Xrayez** [comment]
I confirm, this issue only happens when compiling with `use_lto=yes` while cross-compiling for Windows using MinGW toolchain.

When debugging with `winedebug`, here's the backtrace of the crash:

```
Unhandled exception: 0xc0000029 in 64-bit code (0x000000007bc26288).
0114:fixme:dbghelp:elf_search_auxv can't find symbol in module
Register dump:
 rip:000000007bc26288 rsp:0000000005fa6cb0 rbp:000000007bc64910 eflags:00000202 (   - --  I   - - - )
 rax:000000000000004d rbx:0000000005fa6cd0 rcx:0000000005fa6cd0 rdx:0000000005fa7950
 rsi:0000000005fa7ae0 rdi:0000000005fa7ae0  r8:0000000000000000  r9:0000000005fa6994 r10:0000000005fa6993
 r11:0000000000000293 r12:0000000005fa7950 r13:0000000005fa6e10 r14:0000000005fa6f00 r15:0000000005fafea0
Stack dump:
0x0000000005fa6cb0:  0000000005fa6cd0 000000007bc268ea
0x0000000005fa6cc0:  0000000000000000 0000000005fa6e10
0x0000000005fa6cd0:  00000001c0000029 0000000005fa7950
0x0000000005fa6ce0:  000000007bc26288 0000000000000000
0x0000000005fa6cf0:  0000000005fa6e10 0000000005fa6f00
0x0000000005fa6d00:  0000000005fafea0 000000007bc53046
0x0000000005fa6d10:  0000000000000002 0000000000400000
0x0000000005fa6d20:  000000000093f7c8 0000000000000000
0x0000000005fa6d30:  0000000005fa6f00 0000000005fa6e48
0x0000000005fa6d40:  0000000005fa6e28 0000000000000000
0x0000000005fa6d50:  0000000000000000 0000000000023b70
0x0000000005fa6d60:  0000000005fafea0 0000000005fa7950
Backtrace:
=>0 0x000000007bc26288 EntryPoint+0xffffffffffffffff() in ntdll (0x000000007bc64910)
  1 0x000000007bc5473f EntryPoint+0xffffffffffffffff() in ntdll (0x000000007bc64910)
  2 0x000000007bc5479f EntryPoint+0xffffffffffffffff() in ntdll (0x0000000005fa7a90)
  3 0x0000000006e173da EntryPoint+0xffffffffffffffff() in msvcrt (0x0000000005fa7a90)
  4 0x0000000000b3c188 EntryPoint+0x73aca7() in godot.windows.opt.tools.64.goos (0x0000000005fa7a90)
  5 0x0000000002e7b2c8 EntryPoint+0x2a79de7() in godot.windows.opt.tools.64.goos (0x0000000005fa7a90)
  6 0x000000000093f7c8 EntryPoint+0x53e2e7() in godot.windows.opt.tools.64.goos (0x0000000000941fd0)
  7 0x0000000000942096 EntryPoint+0x540bb5() in godot.windows.opt.tools.64.goos (0x0000000000941fd0)
  8 0x0000000002ef6c33 EntryPoint+0x2af5752() in godot.windows.opt.tools.64.goos (0x0000000000941fd0)
  9 0x0000000001580408 EntryPoint+0x117ef27() in godot.windows.opt.tools.64.goos (0x0000000005fab470)
  10 0x00000000010f8f50 EntryPoint+0xcf7a6f() in godot.windows.opt.tools.64.goos (0x0000000005fab740)
  11 0x00000000010fd1d4 EntryPoint+0xcfbcf3() in godot.windows.opt.tools.64.goos (0x000000000e979f70)
  12 0x0000000001101b95 EntryPoint+0xd006b4() in godot.windows.opt.tools.64.goos (0x00000000038acb54)
  13 0x0000000001102e37 EntryPoint+0xd01956() in godot.windows.opt.tools.64.goos (0x0000000000000000)
  14 0x0000000001bb62f6 EntryPoint+0x17b4e15() in godot.windows.opt.tools.64.goos (0x0000000014ce18f8)
  15 0x0000000001becf10 EntryPoint+0x17eba2f() in godot.windows.opt.tools.64.goos (0x0000000005fabef0)
  16 0x000000000042f095 EntryPoint+0x2dbb4() in godot.windows.opt.tools.64.goos (0x000000000042dd69)
  17 0x00000000004552c2 EntryPoint+0x53de1() in godot.windows.opt.tools.64.goos (0x0000000005fac7a0)
  18 0x0000000000463290 EntryPoint+0x61daf() in godot.windows.opt.tools.64.goos (0x0000000000000004)
  19 0x00000000004013c1 EntryPoint+0xffffffffffffffff() in godot.windows.opt.tools.64.goos (0x0000000000000004)
  20 0x00000000004014f6 EntryPoint+0x15() in godot.windows.opt.tools.64.goos (0x0000000000000000)
  21 0x000000007b62c6e9 EntryPoint+0xffffffffffffffff() in kernel32 (0x0000000000000000)
  22 0x000000007bc588b3 EntryPoint+0xffffffffffffffff() in ntdll (0x0000000000000000)
0x000000007bc26288 EntryPoint+0xffffffffffffffff in ntdll: jmp  0x000000007bc26280 EntryPoint+0xffffffffffffffff
Wine-dbg>pass

```

The backtrace is probably not very useful because apparently it does not contain debug symbols when compiling with LTO. Compiling with `scons target=debug use_lto=yes` produces build errors: godotengine/godot#49212.

LTO stage produces warnings that mention jpegd:

```
thirdparty/zstd/compress/zstd_compress.c: In function 'ZSTD_cParam_getBounds':
thirdparty/zstd/compress/zstd_compress.c:475:1: warning: visibility attribute not supported in this configuration; ignored [-Wattributes]
  475 | }
      | ^
thirdparty/libwebp/src/enc/config_enc.c: In function 'WebPValidateConfig':
thirdparty/libwebp/src/enc/config_enc.c:129:1: warning: visibility attribute not supported in this configuration; ignored [-Wattributes]
  129 | }
      | ^
thirdparty/libwebp/src/enc/picture_csp_enc.c: In function 'WebPPictureHasTransparency':
thirdparty/libwebp/src/enc/picture_csp_enc.c:74:1: warning: visibility attribute not supported in this configuration; ignored [-Wattributes]
   74 | }
      | ^
thirdparty/libwebp/src/enc/picture_csp_enc.c: In function 'WebPPictureImportRGB':
thirdparty/libwebp/src/enc/picture_csp_enc.c:1194:1: warning: visibility attribute not supported in this configuration; ignored [-Wattributes]
 1194 | }
      | ^
thirdparty/libwebp/src/enc/picture_csp_enc.c: In function 'WebPPictureImportRGBA':
thirdparty/libwebp/src/enc/picture_csp_enc.c:1201:1: warning: visibility attribute not supported in this configuration; ignored [-Wattributes]
 1201 | }
      | ^
editor/editor_node.cpp: In member function '__ct_base ':
editor/editor_node.cpp:5652: note: variable tracking size limit exceeded with '-fvar-tracking-assignments', retrying without
 5652 | EditorNode::EditorNode() {
      | 
thirdparty/jpeg-compressor/jpgd.cpp: In member function 'make_huff_table':
thirdparty/jpeg-compressor/jpgd.cpp:2300:15: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
 2300 |   huffsize[p] = 0;
      |               ^
thirdparty/jpeg-compressor/jpgd.cpp:2276:9: note: at offset [0, 2147483646] to object 'huffsize' with size 258 declared here
 2276 |   uint8 huffsize[258];
      |         ^
thirdparty/libwebp/src/enc/picture_enc.c: In function 'WebPMemoryWrite':
thirdparty/libwebp/src/enc/picture_enc.c:216:1: warning: visibility attribute not supported in this configuration; ignored [-Wattributes]
  216 | }
      | ^
thirdparty/libwebp/src/utils/utils.c: In function 'WebPSafeMalloc':
thirdparty/libwebp/src/utils/utils.c:199:1: warning: visibility attribute not supported in this configuration; ignored [-Wattributes]
  199 | }
      | ^
thirdparty/libwebp/src/enc/picture_enc.c: In function 'WebPPictureAlloc':
thirdparty/libwebp/src/enc/picture_enc.c:167:1: warning: visibility attribute not supported in this configuration; ignored [-Wattributes]
  167 | }
      | ^
thirdparty/miniupnpc/miniupnpc/miniupnpc.c: In function 'GetUPNPUrls':
thirdparty/miniupnpc/miniupnpc/miniupnpc.c:496:1: warning: visibility attribute not supported in this configuration; ignored [-Wattributes]
  496 | }
      | ^
thirdparty/miniupnpc/miniupnpc/upnpcommands.c: In function 'UPNP_GetExternalIPAddress':
thirdparty/miniupnpc/miniupnpc/upnpcommands.c:336:1: warning: visibility attribute not supported in this configuration; ignored [-Wattributes]
  336 | }
      | ^
At top level:
lto1: warning: visibility attribute not supported in this configuration; ignored [-Wattributes]
[100%] scons: done building targets.
```
Perhaps this may be a "regression" caused by updating jpgd to 2.00: godotengine/godot#38072.

That said, the exact reason as to why this crash happens in Goost builds and not official Godot builds is still not clear.

As far as I know, official Godot builds use AMD Ryzen processor to compile Godot. I'm using a crappy Intel i3 processor in contrast on my laptop. Maybe that could also explain the difference.

That's my best effort debugging the issue so far, and it looks like it's a build(system) issue. Disabling LTO for jpgd could probably solve this. At some point in time, perhaps official Godot builds would also stumble upon this error under magical circumstances. :slightly_smiling_face:

**Comment by Xrayez** [comment]
I cannot reproduce the crash anymore in Goost 1.1: https://github.com/goostengine/goost/releases/tag/1.1-stable+3.4

I have no idea what fixed it. But for the reference, this time I've built Goost under Windows WSL2 Ubuntu 20.04.3 LTS, not native Fedora 34 host.

But there's a chance that something was fixed in Godot 3.4 and/or updated compiler toolchain as well that automagically resolved this issue.

Closing as fixed! But it would be nice if someone could confirm this as well.
